cartoframes/data/dataset.py

```diff
@@ -1,282 +1,106 @@
-                FROM
-                  taxi_50k
-                ''')
-
-            # show dataset on a map
-            Map(color_continuous_layer(d, 'fare_amount'))
-
-        """
-        return cls(query=query, context=context, state=cls.STATE_REMOTE, is_saved_in_carto=True)
-
-    @classmethod
-    def from_dataframe(cls, df):
-        """Create a Dataset from a local pandas DataFrame.
-
-        Args:
-          df (pandas.DataFrame): pandas DataFrame
-
-        Example:
-
-            Create a Dataset from a pandas Dataframe and then map the data.
-
-            .. code::
-
-                from cartoframes.data import Dataset
-                from cartoframes.viz import Map, Layer
-                import pandas as pd
-
-                df = pd.DataFrame({'lat': [0, 10, 20], 'lng': [20, 10, 0]})
-
-                d = Dataset.from_dataframe(df)
-
-                Map(Layer(d))
-
-        """
-        dataset = cls(df=df, state=cls.STATE_LOCAL)
-        _save_index_as_column(dataset._df)
-        return dataset
-
-    @classmethod
-    def from_geodataframe(cls, gdf):
-        """Create a Dataset from a local GeoPandas GeoDataFrame.
-
-        Args:
-          gdf (geopandas.GeoDataFrame): GeoPandas GeoDataFrame
-
-        Example:
-
-          GeoDataFrame example code taken from `GeoPandas documentation
-          <http://geopandas.org/gallery/create_geopandas_from_pandas.html#creating-a-geodataframe-from-a-dataframe-with-coordinates>`__.
-
-        .. code::
-
-            from cartoframes.data import Dataset
-            from cartoframes.viz import Map, Layer
-            import pandas as pd
-            import geopandas as gpd
-
-            df = pd.DataFrame(
-                {'City': ['Buenos Aires', 'Brasilia', 'Santiago', 'Bogota', 'Caracas'],
-                 'Country': ['Argentina', 'Brazil', 'Chile', 'Colombia', 'Venezuela'],
-                 'Latitude': [-34.58, -15.78, -33.45, 4.60, 10.48],
-                 'Longitude': [-58.66, -47.91, -70.66, -74.08, -66.86]})
-            gdf = gpd.GeoDataFrame(
-                df,
-                geometry=gpd.points_from_xy(df.Longitude, df.Latitude)
-            )
-
-            d = Dataset.from_geodataframe(gdf)
-
-            Map(Layer(d))
-
-        """
-        dataset = cls(df=gdf, state=cls.STATE_LOCAL)
-        _save_index_as_column(dataset._df)
-        return dataset
-
-    @classmethod
-    def from_geojson(cls, geojson):
-        """Create a Dataset from a GeoJSON file (hosted or local).
-
-        Args:
-          gdf (geopandas.GeoDataFrame): GeoPandas GeoDataFrame
-
-        Example:
+    GEOM_TYPE_POINT = GEOM_TYPE_POINT
+    GEOM_TYPE_LINE = GEOM_TYPE_LINE
+    GEOM_TYPE_POLYGON = GEOM_TYPE_POLYGON
+
+    def __init__(self, data, credentials=None, schema=None):
+        self._strategy = self._init_strategy(data, credentials, schema)
+        self._is_saved_in_carto = self._init_saved_in_carto()
+
+    def _init_strategy(self, data, credentials=None, schema=None):
```

Comment:
I'm suggesting an IMO cleaner version:

```suggestion
    def is_geojson_instance(data):
        return isinstance(data, (list, dict)) or (isinstance(data, str) and is_geojson_file_path(data))

    def is_sql_query_instance(data, credentials):
        return isinstance(data, str) and is_sql_query(data)

    def is_table_instance(data, credentials):
        return isinstance(data, str) and not is_sql_query(data)

    def _init_strategy(self, data, credentials=None, schema=None):
        credentials = credentials or _get_default_credentials()

        if isinstance(data, pd.DataFrame):
            return self._getDataFrameDataset(data)
        elif self.is_geojson_instance(data):
            return self._getDataFrameDataset(load_geojson(data))
        elif self.is_sql_query_instance(data, credentials):
            return self._getQueryDataset(data, credentials)
        elif self.is_table_instance(data, credentials):
            return self._getTableDataset(data, credentials, schema)
        else:
            raise ValueError('We can not detect the Dataset type')
```

Reply:
I've been thinking about this and we could take extra advantage of the strategy pattern with a design like this:

1. Let's keep a registry of strategies (it can be a singleton) as somewhere where we (or other developers) could register their `BaseDataset` strategies.

```python

from data.registry import dataset_strategy_registry

dataset_strategy_registry.add(my_custom_dataset_strategy)
```

Each of our `BaseDataset` implementations should add itself to the registry. 

2. Let's open this `_init_strategy` method for extension.

First by adding a `can_work_with(data, credentials)` method to the `BaseDataset`. It will check that `data` and `credentials` are valid for that strategy.

Second let's change the `_init_strategy` method implementation to something like:

```python

def _init_strategy(self, data, credentials=None, schema=None):
  for strategy in dataset_strategy_registry:
    if strategy.can_work_with(data, credentials):
      # return first strategy to be able to deal with `data` and `credentials`
      return strategy.create(data, credentials, schema)

  raise ValueError('We cannot detect the Dataset type')
```

`can_work_with` and `create` need to be part of the abstract class methods and they may be static.

3. This way we can allow others (or ourselves) to add different strategies without having to change the `_init_strategy` implementation.

Let's say we want to support CSV files. We'll just write a new `BaseDataset` implementation which we'll check in `can_work_with` that `data` is a CSV (plus the rest of the abstract class methods), add it to the registry and that's it: CARTOframes will work with CSVs!

This extension model is candidate to be well documented with examples to allow different integrations.

Reply:
AWESOME addition pal! Let me think about it

Reply:
I like the proposal. A lot. 

My concerns about it are:
- The registry should be a singleton. 
- The registry should be accessible anywhere

So, who creates and manages the singleton? We need a new entity for that.

I have some ideas to test, but I am thinking if this is too much for the current PR

Reply:
This situation also applies to `_default_context` and `set_default_context` where we need to import the `_default_context` in the usage instant. It should be resolved in a similar fashion

Reply:
Ok, I have working PoC. I am going to create a new PR over this one

Reply:
Yes, a different PR looks good to me. Thanks for taking it into account :)

Reply:
Done and merged here